### PR TITLE
Adding plugins support to StandardGobblinInstanceDriver

### PR DESCRIPTION
**Pull request by chavdar**
Adding the concept of a plugin which a service that gets started/stopped along with the instance driver and can be used to extend it. For example, a plugin can load a keytab and authenticate with Kerberos.

@pcadabam, @ibuenros can you review?

### CODE REVIEW

File: gobblin-runtime/src/main/java/gobblin/runtime/instance/StandardGobblinInstanceDriver.java [hunk 1]
```diff
@@ -44,28 +49,70 @@
...
+    try {
+      GobblinInstancePlugin plugin = pluginFactory.createPlugin(instance);
+      componentServices.add(plugin);
+      instance.getLog().info("Instantiated plugin: " + plugin);
+      return plugin;
+    }
+    catch (RuntimeException e) {
+      instance.getLog().warn("Failed to create plugin: " + e, e);
+    }
+    return null;
```
Comment (pcadabam) [inline gobblin-runtime/src/main/java/gobblin/runtime/instance/StandardGobblinInstanceDriver.java]: Prefer `Optional` over `null`

File: gobblin-runtime/src/main/java/gobblin/runtime/instance/StandardGobblinInstanceDriver.java [hunk 2]
```diff
@@ -333,6 +387,45 @@ public boolean isInstrumentationEnabled() {
     @Override public void switchMetricContext(MetricContext context) {
       throw new UnsupportedOperationException();
     }
+
+    public List<GobblinInstancePluginFactory> getDefaultPlugins() {
```
Comment (pcadabam) [inline gobblin-runtime/src/main/java/gobblin/runtime/instance/StandardGobblinInstanceDriver.java]: Can you add javadoc?

File: gobblin-runtime/src/main/java/gobblin/runtime/instance/StandardGobblinInstanceDriver.java [hunk 3]
```diff
@@ -333,6 +387,45 @@ public boolean isInstrumentationEnabled() {
...
+      if (!getSysConfig().getConfig().hasPath(PLUGINS_FULL_KEY)) {
+        return Collections.emptyList();
+      }
+
+      String pluginsListStr = getSysConfig().getConfig().getString(PLUGINS_FULL_KEY);
+      List<String> pluginNames = Splitter.on(',').omitEmptyStrings().trimResults().splitToList(pluginsListStr);
```
Comment (pcadabam) [inline gobblin-runtime/src/main/java/gobblin/runtime/instance/StandardGobblinInstanceDriver.java]: Have you looked at `ConfigUtils.getStringList(Config config, String path)` ? It supports both java props style comma separated strings and typesafe style list.

```
Examples below will all return a list [1,2,3] without quotes
a.b=[1,2,3]
a.b=["1","2","3"]
a.b=1,2,3
a.b="1","2","3"
```

File: gobblin-runtime/src/main/java/gobblin/runtime/instance/StandardGobblinInstanceDriver.java [hunk 4]
```diff
@@ -333,6 +387,45 @@ public boolean isInstrumentationEnabled() {
...
+
+      final ClassAliasResolver<GobblinInstancePluginFactory> aliasResolver =
```
Comment (pcadabam) [inline gobblin-runtime/src/main/java/gobblin/runtime/instance/StandardGobblinInstanceDriver.java]: Can you make this an instance variable?

Comment (chavdar) [inline gobblin-runtime/src/main/java/gobblin/runtime/instance/StandardGobblinInstanceDriver.java]: Done.

**Comment by chavdar** [comment]
@ibuenros @pcadabam any more comments?

**Comment by pcadabam** [comment]
LGTM

**Comment by ibuenros** [comment]
Taking a look

**Pull request merged**